### PR TITLE
Add basic reporting core to Shuup (SHUUP-2085)

### PR DESCRIPTION
**Pull request by Pikkupomo**
Add Total Sales report
Add Sales Report
Add Sales per Hour report
Move pdf tools from order printouts to `shuup/utils`
Testing: Cache prices when creating order with product

Refs SHUUP-2085

### CODE REVIEW

File: shuup/reporting/admin_module/__init__.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+from shuup.admin.utils.urls import admin_url
+
+
+class ReportsAdminModule(AdminModule):
+    name = _("Reports")
+    breadcrumbs_menu_entry = MenuEntry(text=name, url="shuup_admin:reporting.list")
+
+    def get_urls(self):
+        return [
+            admin_url("^reporting/$", "shuup.reporting.admin_module.views.ReportView", name="reporting.list")
```
Comment (tulimaki) [inline shuup/reporting/admin_module/__init__.py]: Add permissions

File: shuup/reporting/admin_module/__init__.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        ]
+
+    def get_menu_category_icons(self):
```
Comment (tulimaki) [inline shuup/reporting/admin_module/__init__.py]: This is kind a deprecated already. No need for this in Shuup base at least.

File: shuup/reporting/admin_module/__init__.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        return {self.name: "fa fa-image"}
+
+    def get_menu_entries(self, request):
+        return [
+            MenuEntry(
+                text=self.name,
+                icon="fa fa-image",
+                url="shuup_admin:reporting.list",
+                category=self.name
```
Comment (tulimaki) [inline shuup/reporting/admin_module/__init__.py]: Use `shuup.admin.menu` predefined categories. There should be one for reports.

File: shuup/reporting/admin_module/forms.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+from django import forms
```
Comment (tulimaki) [inline shuup/reporting/admin_module/forms.py]: Missing license header

File: shuup/reporting/admin_module/forms.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+from shuup.reporting.reporter.base import (
+    get_report_class, get_report_classes, get_writer_names
+)
+from shuup.reporting.utils import parse_date_range
+
+
+class ReportForm(forms.Form):
+
+    def __init__(self, **kwargs):
+        super(ReportForm, self).__init__(**kwargs)
```
Comment (tulimaki) [inline shuup/reporting/admin_module/forms.py]: Just a suggestion but would like to see filter based on customer, orderer and creator.

File: shuup/reporting/admin_module/forms.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    def get_extra_filters(self):
+        pass
+
+    def clean(self):
+        data = self.cleaned_data
+        if data.get("date_range") == "custom":
+            try:
+                data["date_range"] = parse_date_range((data["start_date"], data["end_date"]))
+            except Exception as exc:
```
Comment (tulimaki) [inline shuup/reporting/admin_module/forms.py]: Ususally bad idea to just blindly catch all the exceptions.

File: shuup/reporting/reporter/pdf_export.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# -- encoding: UTF-8 --
```
Comment (tulimaki) [inline shuup/reporting/reporter/pdf_export.py]: Missing license header and seems to be duplicate code with `shuup.order_printouts.pdf_export` Maybe this should be somewhere under `shuup.utils.pdf`?

File: shuup/reporting/reports/sales.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+from shuup.reporting.reporter.base import ShuupReportBase
+
+
+class OrderReportMixin(object):
+    def get_objects(self):
+        return Order.objects.filter(
+            shop=self.shop, order_date__range=(self.start_date, self.end_date)).order_by("order_date")
+
+
+class SalesReport(OrderReportMixin, ShuupReportBase):
```
Comment (tulimaki) [inline shuup/reporting/reports/sales.py]: I would like to separate these reports a little maybe even one report per file under reports or reports/sales? I can easily see this file go out of hand.

File: shuup/reporting/admin_module/forms.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                                              help_text=_("For custom date range"))
+        self.fields["end_date"] = DateField(label=_("End Date"), required=False, help_text=_("For custom date range"))
+        self.fields["writer"] = forms.ChoiceField(
+            label=_("Output Format"), initial="html",
+            choices=[(name, name.title()) for name in sorted(get_writer_names())],
+        )
+        self.fields["force_download"] = forms.BooleanField(required=False, label=_("Download"))
+        self.get_extra_filters()
+
+    def get_extra_filters(self):
```
Comment (tulimaki) [inline shuup/reporting/admin_module/forms.py]: Where is this used. Do we really need to have this?

File: shuup/reporting/utils.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        start, end = value[:2]
+    else:
+        raise ValueError("Can't split date range: %r" % value)
+    date_range = (to_datetime(start), to_datetime(end))
+    if any(p is None for p in date_range):
+        raise ValueError("Invalid date range: %r (parsed as %r)" % (value, date_range))
+    return date_range
+
+
+def format_data_with_usd_currency(sales_data):
```
Comment (tulimaki) [inline shuup/reporting/utils.py]: Where is this used? Do we really need this. Should we use `shuup.utils.i18n` for these kind of things?

File: shuup/reporting/utils.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def get_error_data(schema, sales_data):
+    if sales_data.get("data") == "error":
+        return get_empty_data(schema, sales_data, ERROR_MESSAGE)
+    if not sales_data.get("data")[0]:
+        return get_empty_data(schema, sales_data, NO_DATA_MESSAGE)
+
+
+def to_datetime(value):
```
Comment (tulimaki) [inline shuup/reporting/utils.py]: I don't think we need this either. Consider using `shuup.utils.dates:parse_date`

File: shuup/reporting/reports/sales.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
+# -- encoding: UTF-8 --
```
Comment (tulimaki) [inline shuup/reporting/reports/sales.py]: Missing license header...

File: shuup/reporting/locale/fi_FI/LC_MESSAGES/django.po [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
+"Language-Team: LANGUAGE <LL@li.org>\n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+
+msgid "Reports"
+msgstr "Raportit"
+
+#, fuzzy
```
Comment (tulimaki) [inline shuup/reporting/locale/fi_FI/LC_MESSAGES/django.po]: A lot of fuzzines here. I would suggest to leave Finnish translations for Transifex.

File: shuup/reporting/apps.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+# This source code is licensed under the AGPLv3 license found in the
+# LICENSE file in the root directory of this source tree.
+import shuup.apps
+
+
+class AppConfig(shuup.apps.AppConfig):
+    name = "shuup.reporting"
+    provides = {
+        "admin_module": ["shuup.reporting.admin_module:ReportsAdminModule"],
+        "reports": [
```
Comment (tulimaki) [inline shuup/reporting/apps.py]: This new provides is missing documentation.

**Comment by tulimaki** [comment]
I would like to see this splitted a little. Maybe core, admin, and each reports separately. 1k lines in one commit is never a good idea.

**Comment by tulimaki** [comment]
As a suggestion I think better name for this app could be just reports. IMO it would be more in sync with the other apps in Shuup base.

File: shuup/reporting/reports/sales.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+from django.utils.translation import ugettext_lazy as _
+
+from shuup.core.models import Order
+from shuup.core.pricing import TaxfulPrice, TaxlessPrice
+from shuup.reporting.reporter.base import ShuupReportBase
+
+
+class OrderReportMixin(object):
+    def get_objects(self):
+        return Order.objects.filter(
```
Comment (tulimaki) [inline shuup/reporting/reports/sales.py]: This should probably filter cancelled orders or there should be option to remove cancelled orders from the report.

File: shuup/reporting/admin_module/forms.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        if data.get("date_range") == "custom":
+            try:
+                data["date_range"] = parse_date_range((data["start_date"], data["end_date"]))
+            except Exception as exc:
+                self.add_error("__all__", force_text(exc))
+        return data
+
+    def get_report_instance(self):
+        """
+        :rtype: shtats.reporter.base.ReportBase
```
Comment (suutari-ai) [inline shuup/reporting/admin_module/forms.py]: Fix this rtype.

File: shuup/reporting/apps.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+            "shuup.reporting.reports.sales:SalesReport",
+            "shuup.reporting.reports.sales:TotalSales",
+            "shuup.reporting.reports.sales:SalesPerHour",
```
Comment (suutari-ai) [inline shuup/reporting/apps.py]: Maybe the initial reports should be in different app?

File: shuup/reporting/locale/fi_FI/LC_MESSAGES/django.po (hunk 13 above)
Comment (suutari-ai) [inline shuup/reporting/locale/fi_FI/LC_MESSAGES/django.po]: Drop the fi_FI translations if you don't plan to translate them too.

File: shuup/reporting/reporter/base.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+REGISTRY = {}
+
+
+def register_report(slug):
+    def registrator(obj):
+        REGISTRY[slug] = obj
+        obj.slug = slug
+
+    return registrator
```
Comment (suutari-ai) [inline shuup/reporting/reporter/base.py]: Remove the REGISTRY stuff if that is not used any more.

File: shuup/reporting/reporter/base.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+        items[cls.get_name()] = cls
+    return items
+
+
+def get_writer_names():
+    return set(NAME_TO_CLASS)
+
+
+class ShuupReportBase(object):
+    filename_template = None
```
Comment (suutari-ai) [inline shuup/reporting/reporter/base.py]: Maybe rename this to ShuupReportRenderer?

File: shuup/reporting/reporter/base.py [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+    def __init__(self, writer_name, **kwargs):
+        self.writer = NAME_TO_CLASS[writer_name]()
```
Comment (suutari-ai) [inline shuup/reporting/reporter/base.py]: IMHO writer shouldn't be part of reports.

File: shuup/reporting/reporter/writer.py [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+class ExcelReportWriter(ReportWriter):
+    content_type = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
+    extension = ".xlsx"
+
+    def __init__(self):
+        super(ExcelReportWriter, self).__init__()
+        try:
+            import openpyxl
+        except ImportError:
+            raise NotImplementedError("Unable to import openpyxl, is it installed?")
```
Comment (suutari-ai) [inline shuup/reporting/reporter/writer.py]: Move import to file beginning.

File: shuup/reporting/reporter/pdf_export.py (hunk 7 above)
Comment (suutari-ai) [inline shuup/reporting/reporter/pdf_export.py]: Yep, the weasyprint code should be shared with Order Printouts.

File: shuup/reporting/apps.py (hunk 17 above)
Comment (tulimaki) [inline shuup/reporting/apps.py]: I think yes. Would make it easier to replace these initial reports with more advanced reporting.

**Comment by tulimaki** [comment]
retest this please

**Pull request merged**